// ng-brazil/esm5/inscricaoestadual/directive.js
import { __decorate } from "tslib";
import { Directive, forwardRef } from '@angular/core';
import { NG_VALIDATORS } from '@angular/forms';
import { inscricaoestadual } from './validator';
var INSCRICAOESTADUAL_VALIDATOR = {
    provide: NG_VALIDATORS,
    // tslint:disable-next-line: no-use-before-declare
    useExisting: forwardRef(function () { return InscricaoEstadualValidator; }),
    multi: true
};
var InscricaoEstadualValidator = /** @class */ (function () {
    function InscricaoEstadualValidator() {
    }
    InscricaoEstadualValidator.prototype.validate = function (c) {
        return inscricaoestadual('mg')(c);
    };
    InscricaoEstadualValidator = __decorate([
        Directive({
            // tslint:disable-next-line:directive-selector
            selector: '[inscricaoestadual][formControlName],[inscricaoestadual][formControl],[inscricaoestadual][ngModel]',
            providers: [INSCRICAOESTADUAL_VALIDATOR]
        })
    ], InscricaoEstadualValidator);
    return InscricaoEstadualValidator;
}());
export { InscricaoEstadualValidator };
